src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -0,0 +1,206 @@
+        public UrlResolutionTagHelper(IUrlHelper urlHelper, IHtmlEncoder htmlEncoder)
+        {
+            UrlHelper = urlHelper;
+            HtmlEncoder = htmlEncoder;
+        }
+
+        /// <inheritdoc />
+        public override int Order
+        {
+            get
+            {
+                return int.MinValue;
+            }
+        }
+
+        protected IUrlHelper UrlHelper { get; }
+
+        protected IHtmlEncoder HtmlEncoder { get; }
+
+        /// <inheritdoc />
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            IEnumerable<string> attributeNames;
+            if (ElementAttributeLookups.TryGetValue(output.TagName, out attributeNames))
+            {
+                foreach (var attributeName in attributeNames)
+                {
+                    ProcessUrlAttribute(attributeName, output);
+                }
+            }
+
+            // itemid can be present on any HTML element.
+            ProcessUrlAttribute("itemid", output);
+        }
+
+        private void ProcessUrlAttribute(string attributeName, TagHelperOutput output)
+        {
+            IEnumerable<TagHelperAttribute> attributes;
+            if (output.Attributes.TryGetAttributes(attributeName, out attributes))
+            {
+                foreach (var attribute in attributes)
+                {
+                    string resolvedUrl;
+                    if (attribute.Value is string)
+                    {
+                        if (TryResolveUrl(
+                                (string)attribute.Value,
+                                tryEncodeApplicationPath: false,
+                                resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = resolvedUrl;
+                        }
+                    }
+                    else if (attribute.Value is HtmlString)
+                    {
+                        var htmlStringValue = ((HtmlString)attribute.Value).ToString();
+                        if (TryResolveUrl(
+                                htmlStringValue,
+                                tryEncodeApplicationPath: true,
+                                resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = new HtmlString(resolvedUrl);
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool TryResolveUrl(string url, bool tryEncodeApplicationPath, out string resolvedUrl)
+        {
+            resolvedUrl = null;
+
+            if (url == null)
+            {
+                return false;
+            }
+
+            var urlLength = url.Length;
+
+            // Find the start of the potential application relative URL value.
+            var prefixEndIndex = -1;
+            while (++prefixEndIndex < urlLength && char.IsWhiteSpace(url[prefixEndIndex])) ;
```

Comment:
The HTML 5.0 specification says the relevant spaces here are "U+0020 SPACE, "tab" (U+0009), "LF" (U+000A), "FF" (U+000C), and "CR" (U+000D)". For example
- [`script/src`](http://www.w3.org/TR/html5/scripting-1.html#attr-script-src) says
> The value of the attribute must be a valid non-empty URL potentially surrounded by spaces ...

- That links to [text stating](http://www.w3.org/TR/html5/infrastructure.html#valid-non-empty-url-potentially-surrounded-by-spaces)
> A string is a valid non-empty URL potentially surrounded by spaces if, after stripping leading and trailing whitespace from it, it is a valid non-empty URL.

- The "stripping" part links to [text stating](http://www.w3.org/TR/html5/infrastructure.html#strip-leading-and-trailing-whitespace)
> ... the user agent must remove all space characters that are at the start or end of the string.

> "space characters links to the [definition above](http://www.w3.org/TR/html5/infrastructure.html#space-character)

Basically the "whitespace" wording in the middle there is a misdirection.